### PR TITLE
Another try on form reader

### PR DESCRIPTION
**Pull request by pakrym**
Fixes: https://github.com/aspnet/HttpAbstractions/issues/553

Another possible solution to allocations

@Tratcher @halter73 @rynowak @davidfowl

### CODE REVIEW

**Comment by pakrym** [comment]
Before:

After:

**Comment by rynowak** [comment]
You should do profiling in release instead of debug :laughing:

**Comment by rynowak** [comment]
Looks right to me

**Comment by pakrym** [comment]
That's why display class is there. :astonished:

**Comment by rynowak** [comment]
Ya, they are structs in release so they aren't allocated unless you actually go async and need to box.

**Comment by pakrym** [comment]
After in Release:

File: src/Microsoft.AspNetCore.WebUtilities/FormReader.cs [hunk 1]
```diff
@@ -114,49 +134,72 @@ public FormReader(Stream stream, Encoding encoding, ArrayPool<char> charPool)
...
+                if (!TryReadWord('&', ValueLengthLimit, ref _currentValue))
                 {
-                    return word;
+                    return false;
                 }
             }
+            return true;
+        }
+
+        private bool TryReadWord(char seperator, int limit, ref string value)
```
Comment (muratg) [inline src/Microsoft.AspNetCore.WebUtilities/FormReader.cs]: Shouldn't this be `out string value`?

**Comment by muratg** [comment]
This looks good to me. I'll wait a little to see if there are more comments, but if not, I'll put a squirrel here.

File: src/Microsoft.AspNetCore.WebUtilities/FormReader.cs [hunk 2]
```diff
@@ -28,8 +28,12 @@ public class FormReader : IDisposable
         private readonly StringBuilder _builder = new StringBuilder();
         private int _bufferOffset;
         private int _bufferCount;
+        private string _currentKey;
+        private string _currentValue;
+        private bool _eos;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/FormReader.cs]: `_endOfStream`

File: test/Microsoft.AspNetCore.WebUtilities.Tests/FormReaderTests.cs [hunk 3]
```diff
@@ -39,23 +50,23 @@ public class FormReaderTests
...
+            var formCollection = await ReadForm(new FormReader(body));
 
             Assert.Equal("", formCollection["foo"].ToString());
         }
 
         [Theory]
         [InlineData(true)]
         [InlineData(false)]
-        public async Task ReadFormAsync_EmptyValuedWithAdditionalEntryAllowed(bool bufferRequest)
+        public async Task ReadForm_EmptyValuedWithAdditionalEntryAllowed(bool bufferRequest)
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/FormReaderTests.cs]: Why did you rename the tests after an test extension method?

File: test/Microsoft.AspNetCore.WebUtilities.Tests/FormReaderAsyncTest.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    public class FormReaderAsyncTest : FormReaderTests
+    {
+        protected override async Task<Dictionary<string, StringValues>> ReadForm(FormReader reader)
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/FormReaderAsyncTest.cs]: ReadFormAsync

**Comment by davidfowl** [comment]
:shipit:
